src/libraries/sortedLinkedList.se

```diff
@@ -0,0 +1,164 @@
+data comparor
+data collection[](exists, prev, next)
+data count
+data head
+data tail
+data owner
+data initialized
+
+inset('../macros/require.sem')
+
+extern comparisonProviderInterface: [compare:[int256, int256]:int256]
+
+def init():
+    self.owner = msg.sender
+
+def any():
+    require(msg.sender == self.owner or msg.sender == self)
+
+def initialize(comparor: address):
+    require(not self.initialized)
+    self.initialized = 1
+    self.comparor = comparor
+    return(1)
+
+def add(item: int256, hints: arr):
+    require(item != 0)
+
+    if (self.contains(item)):
+        self.remove(item)
+
+    # If this is the only node it is the head / tail and we can early out
+    if (self.count == 0):
+        self.head = item
+        self.tail = item
+        self.collection[item].exists = 1
+        self.count += 1
+        return(1)
+
+    # Use provided hints to get the starting node for traversal
+    nodeValue = self.getBestHint(hints, item)
+
+    # Do traversal to find the insertion point
+    while(self.hasPrev(nodeValue) and self.comparor.compare(nodeValue, item) >= 0):
+        nodeValue = self.getPrev(nodeValue)
```

Comment:
Thinking out loud: I wonder if there is value in creating a function named `findNodeToInsertBefore` which would return the node that `item` should be inserted before and return `0` if `item` should be inserted at head?  I'm still trying to think of a way to make the state update below a bit cleaner.

If such a function existed and was called in place of this while loop (internally it would likely have such a while loop) then below I think we would just need to do something like:
```
# insert at head
if not nodeToInsertBefore:
    self.collection[item].next = 0
    self.collection[item].prev = self.head
    self.collection[self.getPrev(item)].next = item
    self.head = item
# insert not at head
else:
    self.collection[item].next = nodeToInsertBefore
    self.collection[item].prev = self.getPrev(nodeToInsertBefore)
    self.collection[self.getNext(item)].prev = item
    self.collection[self.getPrev(item)].next = item
```
This would remove the need for any comparor usage during the state update, and remove the nested if statements in the current code.

Another potential improvement on this would be to special case head insertion just like you special cased first item insertion:
```
if bestHint == 0: # depends on bestHint returning 0 when head insertion _is_ the best option
    # TODO: insert into head
    return(1)

# TODO: non-head, non-first insertion
```
In this case, we are basically taking out the edge cases early (head insertion and first item insertion) early on leaving the only remaining code path the one where we have a well defined "next" item and we just need to insert before that.  We could even go a step further and have a third early-exit for tail insertion (when `bestHint == self.tail`) which gives us 4 distinct and isolated scenarios.  With some extraction into functions we would end up with code that looks like:

```
if self.count == 0:
    self.insertFirstItem(item)
    return(1)
bestHint = self.getBestHint(hints, item)
if bestHint == 0:
    self.insertAtHead(item)
    return(1)
itemNext = self.findNodeToInsertBefore(item, hint)
if itemNext == self.tail:
    self.insertAtTail(item)
    return(1)
self.insertInMiddle(item, itemNext)
```

Note: The above is purely to help make the code easier to read, it shouldn't have any functional change.  The goal is to make it so the reader can read any one of the helper methods in isolation and make assertions as to whether it does what the function name says.  The reader can then read the above function and assert that "if each of the individual functions do what they say, then the overall algorithm is correct".

Note: If you decide to go down this path, any helper method that mutates state will need to be marked as privet.  There is a macro you can include that has an `assertPrivate()` or something in it.

Note 2: The above is all just style, so I leave it up to you whether to make any changes, or apply a subset of the changes suggested.